flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementSplitter.java

```diff
@@ -0,0 +1,326 @@
+     * @return a map of block name to block statements mappings. The key can be interpreted as name
+     *     of extracted block/method and corresponding List represents individual statements (block'
+     *     lines) for this block.
+     */
+    public Map<String, List<String>> extractBlocks(String context) {
+
+        this.visitor = new BlockStatementVisitor(code, context, parameters);
+        JavaParser javaParser = new JavaParser(visitor.tokenStream);
+        javaParser.getInterpreter().setPredictionMode(PredictionMode.SLL);
+        visitor.visitStatement(javaParser.statement());
+        return visitor.getAllBlocks();
+    }
+
+    /**
+     * Rewrite code block that was used for this object initialization.
+     *
+     * @return a map which key represent rewritten block name and value represents rewritten code
+     *     block, including calls to extracted methods
+     */
+    public Map<String, String> rewriteBlock() {
+        visitor.rewrite();
+        Map<String, String> rewriteBlocks = new HashMap<>();
+        rewriteBlocks.put(visitor.context, visitor.rewriter.getText());
+        return rewriteBlocks;
+    }
+
+    private static void mergeBlocks(
+            Map<String, List<String>> mergeA, Map<String, List<String>> mergeB) {
+
+        for (Entry<String, List<String>> entry : mergeA.entrySet()) {
+            mergeB.merge(
+                    entry.getKey(),
+                    entry.getValue(),
+                    (aBlocks, bBlocks) -> {
+                        List<String> merge = new ArrayList<>(aBlocks.size() + bBlocks.size());
+                        merge.addAll(aBlocks);
+                        merge.addAll(bBlocks);
+                        return merge;
+                    });
+        }
+    }
+
+    private static class BlockStatementVisitor extends JavaParserBaseVisitor<Void> {
+
+        private final List<BlockStatementVisitor> children = new ArrayList<>();
+
+        private final List<ParserRuleContext> extractedSingleBlocks = new ArrayList<>();
+
+        private final String context;
+
+        private final CommonTokenStream tokenStream;
+
+        private final TokenStreamRewriter rewriter;
+
+        private final String parameters;
+
+        private int counter = 0;
+
+        private BlockStatementVisitor(String code, String context, String parameters) {
+            this.tokenStream = new CommonTokenStream(new JavaLexer(CharStreams.fromString(code)));
+            this.rewriter = new TokenStreamRewriter(tokenStream);
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        private BlockStatementVisitor(
+                TokenStreamRewriter rewriter, String context, String parameters) {
+            this.tokenStream = null;
+            this.rewriter = rewriter;
+            this.context = context;
+            this.parameters = parameters;
+        }
+
+        @Override
+        public Void visitStatement(StatementContext ctx) {
```

Comment:
>  My main goal here is to remove addBlocks and mergeBlocks because it's very confusing to understand without proper comments. Readers may think that you're grouping single statements from different levels, but actually you're not.

I see and understand your point here. I will get back to you with this one.

Reply:
@tsreaper 
Sorry for delay on this one.

Regarding:
> I mean something like the following.

I understand your suggestion, I was thinking about similar thing at the beginning of designing the solution.

Let me try to rephrase what I think is main difference in yours and mine proposition.

You are proposing to have single Visitor instance that will recursively parse every statement/block.  Every statement extracted from particular "level" will be added to `blocks` map, where key is "context" that identifies the "level/block" that is currently processed.

In my proposition, I'm creating new Visitor for every new nested "block/level". Every processor maintains its own list of extracted statements. At the end I'm traversing through all visitors and gather all extracted blocks -> mergeBlock method (not the best name I agree).

Reply:
I understand. If it is difficult to refactor this part of code, at least add detailed comments showing that we're only grouping single statements on the same level, not merging different levels together. In `mergeBlocks` you should also assert that two maps do not have the same key.